src/tests/mod.rs

```diff
@@ -17,3 +19,63 @@ macro_rules! test_amber {
 pub fn compile_code<T: Into<String>>(code: T) -> String {
     AmberCompiler::new(code.into(), None).compile().unwrap().1
 }
+
+/// Compile code and compare it with the expected value.
+/// If any difference is found panic and show a friendly diff
+pub fn comp(source: &str, expected: &str) {
+    let mut exp = String::with_capacity(500);
+    let exp = {
+        // Dedent code blocks based on the first line after `r#"`
+        let mut lines = expected.lines();
+        lines.next();
+        let line = lines.next().unwrap_or("");
+        let indent = line.chars().take_while(|&c| c.is_whitespace()).count();
+        exp.push_str(&line[indent..]);
+        exp.push('\n');
+        for line in lines {
+            if line.len() >= indent {
+                exp.push_str(&line[indent..]);
+                exp.push('\n');
+            } else {
+                panic!("Incorrect indentation on line <<<{line}>>>.\nExpected {indent} characters")
+            }
+        }
+        exp.trim_end()
+    };
+
+    let compiled = compile_code(source);
+
+    let header = r#"#!/usr/bin/env bash
+# Written in [Amber](https://amber-lang.com/)
+"#;
+    if !compiled.starts_with(header) {
+        panic!("Unexpected header in generated code {compiled}")
+    }
+    // Ignore newlines before and after the generated code for ease of use
+    let compiled = compiled[65..]
+        .trim_end_matches("\n")
+        .trim_start_matches("\n");
+
+    if compiled == exp {
+        return;
+    };
+
+    let changeset = Changeset::new(&compiled, &exp, " ");
+
+    let mut out = String::new();
+    let mut cnt = 0;
+    for diff in changeset.diffs {
+        match diff {
+            Difference::Same(ref x) => out.push_str(x),
+            Difference::Add(ref x) => out.push_str(&format!("\x1b[32m『{}』\x1b[0m", x)),
+            Difference::Rem(ref x) => out.push_str(&format!("\x1b[31m『{}』\x1b[0m", x)),
+        }
+        cnt += 1;
+    }
+    println!("--- {} differences found ---", cnt);
+    println!("{}", out);
+    println!("--- generated code ---");
+    println!("{}", compiled);
+    println!("=======================================");
+    panic!("Unexpected difference");
```

Comment:
wouldn't this always `panic!` even if there are no differences?

Reply:
also it would print empty reports every time

Reply:
That chunk of code is executed only on test failure: successful test return at line 60:

https://github.com/Ph0enixKM/Amber/pull/240/files#diff-35dd501674c29ffd5ce530e622a8726a658242d3a7fc15c623ad653985ccef46R60